### PR TITLE
VW Slim model prediction.

### PR DESCRIPTION
**Pull request by TomFinley**
In which VW Slim support is added for regression and recommendation output tasks.

### CODE REVIEW

File: src/vw_slim_model/src/output_task.cpp [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+
+   public:
+    IPokerImpl(std::shared_ptr<void> vwModel, std::shared_ptr<void> vwExample);
+    ~IPokerImpl() = default;
+};
+
+OutputTask::IPokerImpl::IPokerImpl(std::shared_ptr<void> vwModel, std::shared_ptr<void> vwExample)
+    : m_model(std::static_pointer_cast<VWModel>(vwModel)), m_example(std::static_pointer_cast<VWExample>(vwExample)) {}
+
+OutputTask::IPoker::IPoker() {}
```
Comment (singlis) [inline src/vw_slim_model/src/output_task.cpp]: The structure seems a little odd and maybe its naming for me, but there is an OutputTask, which makes me think of the C# Task class or at the very least something that I could run. But in reality it implements the IPoker interface, so in the end it acts as an poker and doesnt look like a task. It then implements an internal Poker class that is specific to regression or whatever the algorithm is -- why not just have a RegressionPoker class, RecommendationPoker class, etc?

Comment (TomFinley) [reply src/vw_slim_model/src/output_task.cpp]: > But in reality it implements the IPoker interface

Well, the tasks do not implement the interface, they can return objects that implement this `IPoker` interface. Note also that the entire `IPoker` interface is entirely private. It is not a concept publicly usable by the user. The task is the only thing the user sees. Architecturally, pokers are stateful objects (they own the `m_example` that is built out of the inputs of a particular bound output) whereas the tasks are stateless.

It is analogous, somewhat, to the relationship between `inference::Model` and `inference::ValueUpdater`.

The naming though might need some work. I agree that `Task` is not a great name, though I am not excited about `Poker` being a name. I can explain `OutputTask` fairly easily: it is the task for which you are employing the output of the model. I can't explain `Poker` without getting into a lot of details that the user, in principle, should never have to care about.

So alternatives to task. I guess the idea is, like we have a schema builder for the inputs, we have something that shows out the output is to be interpreted. `OutputInterpreter`? `OutputDescriptor`? Something else? What you think @singlis?

File: src/vw_slim_model/include/vw_slim_model/output_task.hpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace resonance_vw {
+
+/**
+ * @brief An output task describes the outputs relative to the task of a VW model, for example, whether a model's
+ * outputs are for a regression task, classification task, recommendation task, or other.
+ */
+class OutputTask {
+   public:
+    VW_SLIM_MODEL_EXPORT static std::shared_ptr<OutputTask> MakeRegression();
```
[the remark this answers is not in the document]
Comment (TomFinley) [reply src/vw_slim_model/include/vw_slim_model/output_task.hpp]: Yes. Though perhaps, as a separate PR, not this PR? My main objective in this PR was to introduce the machinery necessary to add more tasks. I started with the "easy" one (regression) as we see, but also added the one that was perhaps most useful for our work (recommendation). More should be added, but I was more concerned with getting the mechanics right, at this time.

File: src/vw_slim_model/test/model_test.cpp [hunk 3]
```diff
@@ -8,9 +8,15 @@
 #include <vw_common/schema_builder.hpp>
 #include <vw_slim_model/model.hpp>
 
+#include "data.h"
 #include "env.hpp"
 
+namespace {
+#include "data.h"
```
[the remark this answers is not in the document]
Comment (TomFinley) [reply src/vw_slim_model/test/model_test.cpp]: That is a mistake, thank you @singlis.

File: src/vw_slim_model/src/model.cpp [hunk 4]
```diff
@@ -180,19 +192,21 @@ class StringsStringPeeker final : public ValueUpdaterBase::Peeker<inference::Ten
...
     StringsStringPeeker(std::shared_ptr<inference::IHandle> handle)
-        : ValueUpdaterBase::Peeker<inference::Tensor<std::string>>(handle) {}
+        : ValueUpdater::Peeker<inference::Tensor<std::string>>(handle) {}
 };
 
-std::unique_ptr<ValueUpdaterBase::IPeeker> ValueUpdaterBase::CreatePeeker(
-    priv::SchemaEntry const& entry, std::shared_ptr<vw_slim::example_predict_builder> ex,
-    std::shared_ptr<inference::IHandle> handle) {
+std::unique_ptr<ValueUpdater::IPeeker> ValueUpdater::CreatePeeker(priv::SchemaEntry const& entry,
+                                                                  std::shared_ptr<vw_slim::example_predict_builder> ex,
```
[the remark this answers is not in the document]
Comment (TomFinley) [reply src/vw_slim_model/src/model.cpp]: Oh right. This was before I realized that it was impossible to reuse the builders in this fashion, but I somehow forgot to take it out. Thanks, @singlis. This had some further reaching implications.

File: src/vw_slim_model/src/output_task.cpp (hunk 1 above)
Comment (singlis) [reply src/vw_slim_model/src/output_task.cpp]: Thanks for the explanation @TomFinley. I think initially for me it was weird, but after I saw more how it was used, it makes sense. So please disregard.

**Pull request merged**